README.md
# MidiSheetMusicUnofficial
Unofficial update to the MidiSheetMusic application authored by Madhav Vaidyanathan.

Changes by Justin Vandenberg (2019):
- Updated rendering to use double-buffering to resolve an issue where the logic thread accessed memory owned by the UI thread. This is the primary issue that caused the application to crash following Windows 10 updates
- Changing volume no longer changes system volume
- Volume level is remembered each time the application is opened
- Added toggle for "smooth" scrolling, due to this being a WinForms/GDI application, this is very CPU intensive!


Changelog.txt
MidiSheetMusic 2.7 (Oct 10, 2019)

- Updated rendering to use double-buffering to resolve an issue where the logic thread accessed memory owned by the UI thread. This is the primary issue that caused the application to crash following Windows 10 updates

- Changing volume no longer changes system volume

- Volume level is remembered each time the application is opened

- Added toggle for "smooth" scrolling, due to this being a WinForms/GDI application, this is very CPU intensive!

MidiSheetMusic 2.6 (Jul 28, 2013)

- Fix bug where lyrics are not shown when there's only one track.

- Fix bug where tracks get incorrectly muted.

- Show recently opened files under the File menu.

- Support for drag-and-drop of MIDI files

- Fix several crashes when parsing MIDI files.

- Use MSBuild for command-line build scripts.

MidiSheetMusic 2.5 (Mar 3, 2012)

- New "Save As PDF" feature for Windows version

- Save and restore per-song settings 

- Fix crash that occurs for songs that have notes that are very low or high.

- Allow increasing playback speed, in addition to slowing the playback speed.

- Still play sound for non-displayed tracks.

- Can now click on sheet music to set the starting position.

MidiSheetMusic 2.4 (Apr 16, 2012) 

- Android now supports saving the sheet music as images.

- Added support for displaying lyrics.

- Workaround for wrong TimeSignature format in Acid Music Studio.

- New option to mute tracks, but still keep them displayed.

- Show note letters in english (C, C#, D), Do-Re-Mi, or numbers.

- More easy songs.

- Fixed bug where Save As fails for sample songs.

MidiSheetMusic 2.3 (Jan 15, 2012)

- MidiSheetMusic now ships with over 50 sample MIDI songs.

- Added new menu "Open Sample Song" to select one of the sample songs.

- Added a new menu option "Show Measure Numbers" to display the 
  measure numbers for each staff.

- Added a new menu and dialog "Play Measures in a Loop", which allows you
  to select a range of measures (like 5 through 10), and play just those
  measures in a loop.

- Improved the algorithm that chooses how to connect chords with beams.
  For example, added support for beams connecting 4 chords and 6 chords.

- Fixed a bug where the notes aren't vertically aligned towards the very 
  end of the song.

- Fixed a bug in Windows, where the sound didn't play if the filename was 
  too long.

- Fixed a bug in Windows, where the sound and highlighting didn't sync up
  when the midi file had too many ControlChange events.

- Fix a bug when using horizontal scrolling, where the sheet music 
  suddenly jumps to the start of the song during playback.

- De-select percussion tracks by default.

- Fix a bug on Mac OS X, where you couldn't open midi files with 
  non-ascii filenames.

- Added a project build file (SheetMusicDLL.csproj) for developers who 
  want to build the SheetMusic Control as a standalone DLL library,
  independent of the MidiSheetMusic executable.

- Ported MidiSheetMusic to the Android platform.

MidiSheetMusic 2.2 (June 26, 2011)

- Improved the appearance of the sheet music by doing "full justification".  
  That is, increase the horizontal spacing of the symbols so that they
  fill out the staff.

- Other minor improvements to the sheet music appearance.

- Improved the appearance of the piano, to look more realistic.

- Added a new menu option "Show Note Letters" which displays the actual
  note letters (A, A#, Bb, etc) beside the notes in the sheet music.

- When the song has exactly two tracks, use two different colors for shading
  the notes during playback (one for left hand, one for right hand).

- In the piano, keep the notes shaded for their full duration.

- Added a new menu option "Show Note Letters" which displays the actual
  note letters (A, A#, Bb, etc) beside the notes in the sheet music.

- Fixed a bug with the G-flat key signature, where B-flat was being 
  displayed as A.

- Fixed a bug in the "Choose Instruments" dialog, where typing in your own
  instrument (rather than choosing from the drop-down list) causes the
  application to crash.

- Retain the pause position while changing the sheet music settings.
  Previously, changing any settings (like transpose) while the music was
  paused caused the music to restart from the beginning.  Now, the music
  will start from where it was paused.

- Print the page numbers at the bottom right-hand corner.

- Added tutorial videos to the website.


MidiSheetMusic 2.1 (Feb 6, 2011)

- Ported MidiSheetMusic to Ubuntu Linux, using C#/Mono.
  It uses Timidity for playing the MIDI sound.

- Changed how we choose black notes.  Before, we would use either all
  sharps or all flats, depending on the key signature.  If the key
  was F major (1 flat), then we only use flats (e.g. we use G-flat
  instead of F-sharp).  Now, we use the following method:

  * Use F# and C# instead of G-flat and D-flat
  * Use B-flat instead of A#
  * For all other black keys, use the same accidental (sharp/flat)
    that the key signature uses. For C major, use sharps.

  This required modifications to the KeySignature methods
  GetWhiteNote and GetAccidental.

- Changed MidiFile.ReadTrack so that if it encounters a truncated
  Midi file, it returns what it has parsed so far, instead of 
  throwing an exception.

- Improved the accuracy and speed of the MidiFile.SplitTrack() method.

- Previously, the Volume slider only worked on Windows XP.
  Fixed it to work on Windows Vista and Windows 7.

- When printing, include the MIDI filename at the top of the first page.

- Fixed errors in the MusicSymbolDiagram at the bottom of README.html.

- Display the message "Use the menu File:Open to select a MIDI file" on
  startup.  


MidiSheetMusic 2.0 (Oct 10, 2010)

- Added the following new features:
  * Playing the Midi sound
  * Changing the Midi sound options, such as speed, transposed notes,
    instruments, and tracks to play.
  * Highlighting the sheet music notes as the midi sound is being played.

Classes/Volume.cs
/*
 * Copyright (c) 2009-2011 Madhav Vaidyanathan
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

using System;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.InteropServices;

/** @class Volume
 *  This volume class has one simple static method:
 *  void SetVolume(int value)
 *  where value is between 0 and 100.
 *
 */
namespace MidiSheetMusic
{


    /* Windows XP uses the MIXERCONTROL structs to set the Volume */
    public struct MIXERCONTROLDETAILS
    {
        public int cbStruct;
        public int dwControlID;
        public int cChannels;
        public int item;
        public int cbDetails;
        public IntPtr paDetails;
    }

    public struct MIXERCONTROLDETAILS_UNSIGNED
    {
        public int dwValue;
    }


    /* Windows Vista uses several COM interfaces to set the Volume. */

    [Guid("5CDF2C82-841E-4546-9722-0CF74078229A"),
     InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]

    public interface IAudioEndpointVolume
    {
        int RegisterControlChangeNotify(IntPtr p);
        int UnregisterControlChangeNotify(IntPtr p);
        int GetChannelCount(ref uint channelCount);
        int SetMasterVolumeLevel(float levelDB, Guid context);
        int SetMasterVolumeLevelScalar(float level, Guid context);
        int GetMasterVolumeLevel(ref float levelDB);
        int GetMasterVolumeLevelScalar(ref float level);
        int SetChannelVolumeLevel(uint channel, float levelDB, Guid context);
        int SetChannelVolumeLevelScalar(uint channel, float level, Guid context);
        int GetChannelVolumeLevel(uint channel, ref float leveldb);
        int GetChannelVolumeLevelScalar(uint channel, ref float level);
        int SetMute(bool mute, Guid context);
        int GetMute(ref bool mute);
        int GetVolumeStepInfo(ref uint step, ref uint stepCount);
        int VolumeStepUp(Guid context);
        int VolumeStepDown(Guid context);
        int QueryHardwareSupport(ref uint mask);
        int GetVolumeRange(ref float min, ref float max, ref float inc);
    }

    [Guid("D666063F-1587-4E43-81F1-B948E807363F"),
     InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface IMMDevice
    {
        int Activate(ref Guid id, uint context, IntPtr p, ref IntPtr endpoint);
        int OpenPropertyStore(int access, ref IntPtr properties);
        int GetId(ref string id);
        int GetState(ref int state);
    }


    [Guid("A95664D2-9614-4F35-A746-DE8DB63617E6"),
     InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    public interface IMMDeviceEnumerator
    {
        int EnumAudioEndpoints(int dataFlow, int dwStateMask, ref IntPtr devices);
        int GetDefaultAudioEndpoint(int dataFlow, int role, ref IntPtr device);
        int GetDevice(string id, ref IntPtr device);
        int RegisterEndpointNotificationCallback(IntPtr client);
        int UnregisterEndpointNotificationCallback(IntPtr client);
    }


    public class Volume
    {
        public const int MIXER_SETCONTROLDETAILSF_VALUE = 0x0;

        /* Win32 API audio function prototypes  */
        [DllImport("winmm.dll")]
        private static extern int mixerOpen(out int mix, int mixerId, int callback, int instance, int flags);

        [DllImport("winmm.dll")]
        private static extern int mixerClose(int handle);

        [DllImport("winmm.dll")]
        private static extern int
        mixerSetControlDetails(int handle, ref MIXERCONTROLDETAILS details, int fdwDetails);

        [DllImport("ole32.dll")]
        private static extern int
        CoCreateInstance(ref Guid id, [MarshalAs(UnmanagedType.IUnknown)] object inner,
                         int context, ref Guid iid,
                         [MarshalAs(UnmanagedType.IUnknown)] out object comObject);

        [DllImport("ole32.dll")]
        private static extern int CoInitialize(object arg);


        /* Set the Volume on Windows Vista by obtaining an
         * IAudioEndpointVolume COM object. 
         */
        public static void SetVolumeWindowsVista(int value)
        {
            object enumerator = null;
            Guid devEnumeratorGuid = new Guid("BCDE0395-E52F-467C-8E3D-C4579291692E");
            Guid idevEnumeratorGuid = new Guid("00000000-0000-0000-C000-000000000046");
            int CLSCTX_INPROC_SERVER = 1;

            CoInitialize(null);
            int result = CoCreateInstance(ref devEnumeratorGuid, null,
                                           CLSCTX_INPROC_SERVER, ref idevEnumeratorGuid,
                                           out enumerator);
            if (result != 0 || enumerator == null)
            {
                return;
            }
            IMMDeviceEnumerator devEnum = enumerator as IMMDeviceEnumerator;
            if (devEnum == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                return;
            }
            IntPtr devicePtr = IntPtr.Zero;
            int eRender = 0; int eConsole = 0;
            int ret = devEnum.GetDefaultAudioEndpoint(eRender, eConsole, ref devicePtr);
            if (ret != 0)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                return;
            }
            object obj = System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(devicePtr);
            IMMDevice device = obj as IMMDevice;
            if (device == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
            }
            Guid iAudioEndPointVolGuid = new Guid("5CDF2C82-841E-4546-9722-0CF74078229A");
            uint context = (uint)0x17;
            IntPtr activationParams = IntPtr.Zero;
            IntPtr endPoint = IntPtr.Zero;
            ret = device.Activate(ref iAudioEndPointVolGuid, context,
                                  activationParams, ref endPoint);
            if (ret != 0)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
                return;
            }
            obj = System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(endPoint);
            IAudioEndpointVolume audioEndpoint = obj as IAudioEndpointVolume;
            if (audioEndpoint == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
                return;
            }
            Guid empty = Guid.Empty;
            audioEndpoint.SetMasterVolumeLevelScalar((float)(value / 100.0), empty);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(audioEndpoint);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
        }


        /* Set the SW Synth volume on Windows XP, by using the
         * Win32 API function mixerSetControlDetails().
         */
        public static void SetVolumeWindowsXP(int value)
        {
            /* The actual master volume range is 0 to 65536. Scale the value */
            value = (int)(value * 65536 / 100.0);

            int mixerHandle;
            int ret = mixerOpen(out mixerHandle, 0, 0, 0, 0);

            MIXERCONTROLDETAILS details = new MIXERCONTROLDETAILS();
            MIXERCONTROLDETAILS_UNSIGNED volumeDetails = new MIXERCONTROLDETAILS_UNSIGNED();
            details.item = 0;

            /* The ControlID is which volume to modify:
             *   Master    = 0
             *   Wave      = 2
             *   SW Synth  = 4
             *   CD Player = 6
             */
            details.dwControlID = 4;
            details.cbStruct = Marshal.SizeOf(details);
            details.cbDetails = Marshal.SizeOf(volumeDetails);
            details.cChannels = 1;
            volumeDetails.dwValue = value;

            details.paDetails = Marshal.AllocCoTaskMem(Marshal.SizeOf(typeof(MIXERCONTROLDETAILS_UNSIGNED)));
            Marshal.StructureToPtr(volumeDetails, details.paDetails, false);
            ret = mixerSetControlDetails(mixerHandle, ref details, MIXER_SETCONTROLDETAILSF_VALUE);
            mixerClose(mixerHandle);
        }

        /* Set the Volume on Linux by using the command:
         * # amixer -c 0 sset Master playback 80%
         * amixer is the command-line mixer for the ALSA driver, found
         * in package alsa-utils on Ubuntu.
         */
        public static void SetVolumeLinux(int value)
        {
            int dbvalue = (int)(Math.Log10(value) * 100.0 / 2.0);
            if (value <= 1)
            {
                dbvalue = 0;
            }
            try
            {
                ProcessStartInfo info = new ProcessStartInfo();
                info.CreateNoWindow = true;
                info.RedirectStandardOutput = true;
                info.RedirectStandardInput = true;
                info.UseShellExecute = false;
                info.FileName = "/usr/bin/amixer";
                info.Arguments = " -c 0 sset Master playback " + dbvalue + "%";
                Process amixer = new Process();
                amixer.StartInfo = info;
                amixer.Start();
                amixer.WaitForExit();
                amixer.Close();
            }
            catch (Exception e)
            {
            }
        }

        public static void SetVolume(int value)
        {
            if (Type.GetType("Mono.Runtime") != null)
            {
                SetVolumeLinux(value);
            }
            else if (Environment.OSVersion.Version.Major == 5)
            {
                SetVolumeWindowsXP(value);
            }
            else
            {
                SetVolumeWindowsVista(value);
            }
        }

        /// <summary>
        /// Gets the current system volume.
        /// </summary>
        /// <returns>The current system volume, range 0-100.</returns>
        public static int GetVolume()
        {
            if (Type.GetType("Mono.Runtime") != null)
            {
                // JV: I have no plans to personally use this on Linux, simply use
                // the existing default of 100
                return 100;
            }
            else if (Environment.OSVersion.Version.Major == 5)
            {
                // JV: WinXP is not supported by Microsoft, I won't either
                return 100;
            }
            else
            {
                // JV: This should prevent anyone from losing their hearing.
                // The application used to set the system volume to 100 on start,
                // which would be really bad for headphone and laptop users
                return GetVolumeWindowsVista();
            }
        }

        /// <summary>
        /// Get the current system volume, if an error is encountered return the default 100.
        /// </summary>
        /// <returns></returns>
        private static int GetVolumeWindowsVista()
        {
            object enumerator = null;
            Guid devEnumeratorGuid = new Guid("BCDE0395-E52F-467C-8E3D-C4579291692E");
            Guid idevEnumeratorGuid = new Guid("00000000-0000-0000-C000-000000000046");
            int CLSCTX_INPROC_SERVER = 1;

            CoInitialize(null);
            int result = CoCreateInstance(ref devEnumeratorGuid, null,
                                           CLSCTX_INPROC_SERVER, ref idevEnumeratorGuid,
                                           out enumerator);
            if (result != 0 || enumerator == null)
            {
                return 100;
            }
            IMMDeviceEnumerator devEnum = enumerator as IMMDeviceEnumerator;
            if (devEnum == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                return 100;
            }
            IntPtr devicePtr = IntPtr.Zero;
            int eRender = 0; int eConsole = 0;
            int ret = devEnum.GetDefaultAudioEndpoint(eRender, eConsole, ref devicePtr);
            if (ret != 0)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                return 100;
            }
            object obj = System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(devicePtr);
            IMMDevice device = obj as IMMDevice;
            if (device == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
            }
            Guid iAudioEndPointVolGuid = new Guid("5CDF2C82-841E-4546-9722-0CF74078229A");
            uint context = (uint)0x17;
            IntPtr activationParams = IntPtr.Zero;
            IntPtr endPoint = IntPtr.Zero;
            ret = device.Activate(ref iAudioEndPointVolGuid, context,
                                  activationParams, ref endPoint);
            if (ret != 0)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
                return 100;
            }
            obj = System.Runtime.InteropServices.Marshal.GetObjectForIUnknown(endPoint);
            IAudioEndpointVolume audioEndpoint = obj as IAudioEndpointVolume;
            if (audioEndpoint == null)
            {
                System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);
                System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
                return 100;
            }

            float volume = 0.0f;
            audioEndpoint.GetMasterVolumeLevelScalar(ref volume);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(audioEndpoint);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(device);
            System.Runtime.InteropServices.Marshal.ReleaseComObject(devEnum);

            return (int)(volume * 100f);
        }
    }

}





Classes/MidiOptions.cs
/*
 * Copyright (c) 2007-2013 Madhav Vaidyanathan
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 */

using System;
using System.IO;
using System.Collections.Generic;
using System.Text;
using System.Drawing;

namespace MidiSheetMusic
{


    /// <summary>
    /// The MidiOptions class contains the available options for
    /// modifying the sheet music and sound.These options are
    /// collected from the menu/dialog settings, and then are passed
    /// to the SheetMusic and MidiPlayer classes.
    /// </summary>
    public class MidiOptions
    {

        // The possible values for showNoteLetters
        public const int NoteNameNone = 0;
        public const int NoteNameLetter = 1;
        public const int NoteNameFixedDoReMi = 2;
        public const int NoteNameMovableDoReMi = 3;
        public const int NoteNameFixedNumber = 4;
        public const int NoteNameMovableNumber = 5;

        // Sheet Music Options
        public string filename;         // The full Midi filename
        public string title;            // The Midi song title
        public bool[] tracks;           // Which tracks to display (true = display)
        public bool scrollVert;         // Whether to scroll vertically or horizontally
        public bool smoothScroll;       // JV: Whether to scroll smoothly (large performance impact)
        public bool largeNoteSize;      // Display large or small note sizes
        public bool twoStaffs;          // Combine tracks into two staffs?
        public int showNoteLetters;     // Show the name (A, A#, etc) next to the notes
        public bool showLyrics;         // Show the lyrics under each note
        public bool showMeasures;       // Show the measure numbers for each staff
        public int shifttime;           // Shift note starttimes by the given amount
        public int transpose;           // Shift note key up/down by given amount
        public int key;                 // Use the given KeySignature (notescale)
        public TimeSignature time;      // Use the given time signature
        public int combineInterval;     // Combine notes within given time interval (msec)
        public Color[] colors;          // The note colors to use
        public Color shadeColor;        // The color to use for shading.
        public Color shade2Color;       // The color to use for shading the left hand piano


        // Sound options
        public bool[] mute;                 // Which tracks to mute (true = mute)
        public int tempo;                   // The tempo, in microseconds per quarter note
        public int pauseTime;               // Start the midi music at the given pause time
        public int[] instruments;           // The instruments to use per track
        public bool useDefaultInstruments;  // If true, don't change instruments
        public bool playMeasuresInLoop;     // Play the selected measures in a loop
        public int playMeasuresInLoopStart; // Start measure to play in loop
        public int playMeasuresInLoopEnd;   // End measure to play in loop


        public MidiOptions()
        {
        }

        public MidiOptions(MidiFile midifile)
        {
            filename = midifile.FileName;
            title = Path.GetFileName(midifile.FileName);
            int numtracks = midifile.Tracks.Count;
            tracks = new bool[numtracks];
            mute = new bool[numtracks];
            instruments = new int[numtracks];
            for (int i = 0; i < tracks.Length; i++)
            {
                tracks[i] = true;
                mute[i] = false;
                instruments[i] = midifile.Tracks[i].Instrument;
                if (midifile.Tracks[i].InstrumentName == "Percussion")
                {
                    tracks[i] = false;
                    mute[i] = true;
                }
            }
            useDefaultInstruments = true;
            scrollVert = true;
            smoothScroll = true;
            largeNoteSize = false;
            if (tracks.Length == 1)
            {
                twoStaffs = true;
            }
            else
            {
                twoStaffs = false;
            }
            showNoteLetters = NoteNameNone;
            showLyrics = true;
            showMeasures = false;
            shifttime = 0;
            transpose = 0;
            key = -1;
            time = midifile.Time;
            colors = null;
            shadeColor = Color.FromArgb(210, 205, 220);
            shade2Color = Color.FromArgb(80, 100, 250);
            combineInterval = 40;
            tempo = midifile.Time.Tempo;
            pauseTime = 0;
            playMeasuresInLoop = false;
            playMeasuresInLoopStart = 0;
            playMeasuresInLoopEnd = midifile.EndTime() / midifile.Time.Measure;
        }

        /* Join the array into a comma separated string */
        static string Join(bool[] values)
        {
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < values.Length; i++)
            {
                if (i > 0)
                {
                    result.Append(",");
                }
                result.Append(values[i].ToString());
            }
            return result.ToString();
        }

        static string Join(int[] values)
        {
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < values.Length; i++)
            {
                if (i > 0)
                {
                    result.Append(",");
                }
                result.Append(values[i].ToString());
            }
            return result.ToString();
        }

        static string Join(Color[] values)
        {
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < values.Length; i++)
            {
                if (i > 0)
                {
                    result.Append(",");
                }
                result.Append(ColorToString(values[i]));
            }
            return result.ToString();
        }

        static string ColorToString(Color c)
        {
            return "" + c.R + " " + c.G + " " + c.B;
        }

        /* Convert this MidiOptions object into a INI formatted section.
         * [title]
         * filename=C:/path/to/file.mid
         * version=2.5
         * tracks=true,false,true
         * mute=true,false,true
         * instruments=24,1,52
         * useDefaultInstruments=true
         * time=3,4,720,10000    // numerator,denominator,quarter,tempo
         * scrollVert=true
         * smoothScroll=true
         * largeNoteSize=false
         * showLyrics=true
         * twoStaffs=true
         * showNoteLetters=true
         * transpose=-2
         * key=1               // Key signature (NoteScale.N)
         * combineInterval=80
         * showMeasures=true
         * playMeasuresInLoop=false
         * playMeasuresInLoopStart=2
         * playMeasuresInLoopEnd=10
         * shadeColor=244 159 24   // R G B
         * shade2Color=244 159 24   // R G B
         * colors=244 159 24, 253 92 143, ...
         */
        public SectionINI ToINI()
        {
            SectionINI section = new SectionINI();
            section.Section = title;
            try
            {
                section.Properties["filename"] = filename;
                section.Properties["version"] = "2.6.0";
                section.Properties["tracks"] = Join(tracks);
                section.Properties["mute"] = Join(mute);
                section.Properties["instruments"] = Join(instruments);
                section.Properties["useDefaultInstruments"] = useDefaultInstruments.ToString();
                if (time != null)
                {
                    int[] values = { time.Numerator, time.Denominator, time.Quarter, time.Tempo };
                    section.Properties["time"] = Join(values);
                }
                section.Properties["scrollVert"] = scrollVert.ToString();
                section.Properties["smoothScroll"] = smoothScroll.ToString();
                section.Properties["largeNoteSize"] = largeNoteSize.ToString();
                section.Properties["showLyrics"] = showLyrics.ToString();
                section.Properties["twoStaffs"] = twoStaffs.ToString();
                section.Properties["showNoteLetters"] = showNoteLetters.ToString();
                section.Properties["transpose"] = transpose.ToString();
                section.Properties["key"] = key.ToString();
                section.Properties["combineInterval"] = combineInterval.ToString();
                section.Properties["shadeColor"] = ColorToString(shadeColor);
                section.Properties["shade2Color"] = ColorToString(shade2Color);
                if (colors != null)
                {
                    section.Properties["colors"] = Join(colors);
                }
                section.Properties["showMeasures"] = showMeasures.ToString();
                section.Properties["playMeasuresInLoop"] = playMeasuresInLoop.ToString();
                section.Properties["playMeasuresInLoopStart"] = playMeasuresInLoopStart.ToString();
                section.Properties["playMeasuresInLoopEnd"] = playMeasuresInLoopEnd.ToString();
            }
            catch (Exception e)
            {
            }
            return section;
        }


        /* Initialize a MidiOptions from the section properties of an INI text file */
        public MidiOptions(SectionINI section)
        {
            title = section.Section;
            filename = section.GetString("filename");
            tracks = section.GetBoolArray("tracks");
            mute = section.GetBoolArray("mute");
            if (mute != null && section.Properties["version"] == "2.5.0")
            {
                // MidiSheetMusic 2.5 stored the mute value incorrectly
                for (int i = 0; i < mute.Length; i++)
                {
                    mute[i] = false;
                }
            }

            instruments = section.GetIntArray("instruments");
            int[] timesig = section.GetIntArray("time");
            if (timesig != null && timesig.Length == 4)
            {
                time = new TimeSignature(timesig[0], timesig[1], timesig[2], timesig[3]);
            }
            useDefaultInstruments = section.GetBool("useDefaultInstruments");
            scrollVert = section.GetBool("scrollVert");
            smoothScroll = section.GetBool("smoothScroll");
            largeNoteSize = section.GetBool("largeNoteSize");
            showLyrics = section.GetBool("showLyrics");
            twoStaffs = section.GetBool("twoStaffs");
            showNoteLetters = section.GetInt("showNoteLetters");
            transpose = section.GetInt("transpose");
            key = section.GetInt("key");
            combineInterval = section.GetInt("combineInterval");
            showMeasures = section.GetBool("showMeasures");
            playMeasuresInLoop = section.GetBool("playMeasuresInLoop");
            playMeasuresInLoopStart = section.GetInt("playMeasuresInLoopStart");
            playMeasuresInLoopEnd = section.GetInt("playMeasuresInLoopEnd");

            Color color = section.GetColor("shadeColor");
            if (color != Color.White)
            {
                shadeColor = color;
            }
            color = section.GetColor("shade2Color");
            if (color != Color.White)
            {
                shade2Color = color;
            }
            colors = section.GetColorArray("colors");
        }


        /* Merge in the saved options to this MidiOptions.*/
        public void Merge(MidiOptions saved)
        {
            if (saved.tracks != null && saved.tracks.Length == tracks.Length)
            {
                for (int i = 0; i < tracks.Length; i++)
                {
                    tracks[i] = saved.tracks[i];
                }
            }
            if (saved.mute != null && saved.mute.Length == mute.Length)
            {
                for (int i = 0; i < mute.Length; i++)
                {
                    mute[i] = saved.mute[i];
                }
            }
            if (saved.instruments != null && saved.instruments.Length == instruments.Length)
            {
                for (int i = 0; i < instruments.Length; i++)
                {
                    instruments[i] = saved.instruments[i];
                }
            }
            if (saved.time != null)
            {
                time = new TimeSignature(saved.time.Numerator, saved.time.Denominator,
                        saved.time.Quarter, saved.time.Tempo);
            }
            useDefaultInstruments = saved.useDefaultInstruments;
            scrollVert = saved.scrollVert;
            smoothScroll = saved.smoothScroll;
            largeNoteSize = saved.largeNoteSize;
            showLyrics = saved.showLyrics;
            twoStaffs = saved.twoStaffs;
            showNoteLetters = saved.showNoteLetters;
            transpose = saved.transpose;
            key = saved.key;
            combineInterval = saved.combineInterval;
            if (saved.shadeColor != Color.White)
            {
                shadeColor = saved.shadeColor;
            }
            if (saved.shade2Color != Color.White)
            {
                shade2Color = saved.shade2Color;
            }
            if (saved.colors != null)
            {
                colors = saved.colors;
            }
            showMeasures = saved.showMeasures;
            playMeasuresInLoop = saved.playMeasuresInLoop;
            playMeasuresInLoopStart = saved.playMeasuresInLoopStart;
            playMeasuresInLoopEnd = saved.playMeasuresInLoopEnd;
        }
    }

}


